kotlinx-coroutines-core/common/src/channels/ArrayChannel.kt

```diff
@@ -40,7 +40,7 @@ internal open class ArrayChannel<E>(
     private var size: Int = 0 // Invariant: size <= capacity
 
     protected final override val isBufferAlwaysEmpty: Boolean get() = false
-    protected final override val isBufferEmpty: Boolean get() = size == 0
+    protected final override val isBufferEmpty: Boolean get() = lock.withLock { size == 0 }
```

Comment:
I would not be so scared since
1. this is actually not a _fast path_ because it suspends and suspensions are expensive
2. lock-based channel implementations in Golang are not worse in practice
3. this code will be completely removed with the new fast and scalable channels
4. this extra lock acquisition does not affect uncontended scenarios 

At the same time, your suggestion breaks the semantics of `isBufferEmpty` and makes the code harder to understand.

Reply:
>  Moreover, it looks scary, since this is an atomic multi-word operation that other threads might be helping with.
This change adds only one read under the lock -- seems like a safe change